server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java

```diff
@@ -0,0 +1,344 @@
+            lua.setTable(-3);
+         }
+         // push the "map" table to the "outer" table
+         lua.setTable(-3);
+         lua.push("len");
+         lua.push(map.size());
+         // push the "len" item to the "outer" table
+         lua.setTable(-3);
+      }
+   }
+
+   @Override
+   public void map(Map<?, ?> map, SerializationHint.KeyValueHint keyValueHint) {
+      if (map == null) {
+         nulls();
+      } else {
+         /*
+          * A map is represented as a lua table with two elements: 'map', containing a table
+          * with the map entries, and `len` containing the size of the set.
+          */
+         lua.checkStack(3);
+         // create the "outer" table
+         lua.newTable();
+         lua.push("map");
+         // create the actual "map" table
+         lua.newTable();
+         for (Map.Entry<?, ?> entry : map.entrySet()) {
+            // push the key
+            keyValueHint.key().serialize(entry.getKey(), this);
+            // push the value
+            keyValueHint.value().serialize(entry.getValue(), this);
+            // push the top two elements on the stack (key at -2 and value at -1) to the table (at -3)
+            lua.setTable(-3);
+         }
+         // push the "map" table to the "outer" table
+         lua.setTable(-3);
+         lua.push("len");
+         lua.push(map.size());
+         // push the "len" item to the "outer" table
+         lua.setTable(-3);
+      }
+   }
+
+   @Override
+   public void error(CharSequence value) {
+      lua.checkStack(3);
+      // create a table with the "err" entry
+      luaPushError(lua, value.toString());
+      // add `ignore_error_stats_update = true` to the table
+      lua.push("ignore_error_stats_update");
+      lua.push(true);
+      lua.setTable(-3);
+   }
+
+   @Override
+   public void error(Throwable t) {
+      error(filterCause(t).getMessage());
+   }
+
+   @Override
+   public <T> void write(T object, JavaObjectSerializer<T> serializer) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
+   }
+
+   @Override
+   public <T> void write(JavaObjectSerializer<T> serializer) {
+      serializer.accept(null, this);
+   }
+
+   @Override
+   public void serialize(Object object) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
+   }
+
+   @Override
+   public void writeNumericPrefix(byte symbol, long number) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
+   }
+
+   @Override
+   public void arrayStart(int size) {
+      assert arrayIndex == 0;
+      lua.checkStack(2);
+      lua.newTable();
+      arrayIndex = 1;
+   }
+
+   @Override
+   public void arrayEnd() {
+      if (arrayIndex > 1) {
+         // expects the stack to contain an array index (at -2) and a value (at -1) and adds them to the array table
+         lua.setTable(-3);
+      }
+      arrayIndex = 0;
+   }
+
+   @Override
+   public void arrayNext() {
```

Comment:
Have to admit, I am not a fan of this method from the caller perspective as I feel we could know we are inside an array and do this. But it does make this class a lot less cluttered, so I will not push back on it.